File: chess_cpp/statecheckers/PieceGoneChecker.cpp
//
// Created by Hrach Khachatryan on 12/16/19.
//

#include "PieceGoneChecker.h"
#include "../piecies/Pawn.h"

using namespace std;

template<typename Base, typename T>
inline bool instanceof(const T *) {
    return is_base_of<Base, T>::value;
}

void check(Board *board, Spot *start, Spot *end, vector<string> messages) {
    if (end != nullptr && end->getPiece().get() != nullptr) {
        if (instanceof<Pawn>(end->getPiece().get()))
            messages.push_back("Pawn is gone");
        else
            messages.push_back("Piece is gone");

    }
}

File: oop_chess/src/chess/statecheckers/PieceGoneChecker.java
package chess.statecheckers;

import chess.Board;
import chess.Spot;
import chess.piecies.Pawn;

import java.util.List;

public class PieceGoneChecker extends BaseStateChangeChecker {

    @Override
    public void check(Board board, Spot start, Spot end, List<String> messages) {
        if(end != null && end.getPiece() != null) {
            if(end.getPiece() instanceof Pawn)
                messages.add("Pawn is gone");
            else
                 messages.add("Piece is gone");

        }

    }
}
